### DISCUSSION TITLE
Drastic speed difference between GCS and S3

### DISCUSSION THREAD

**Discussion by mpiannucci** [fsspec/kerchunk#395] [Q&A]
I am not sure if this is an fsspec issue or a kerchunk one (probably the former) but I am posting here because it is affecting the kerchunking workflow. 

I have the same kerchunked dataset created pointing at s3 and gcs: 

s3: `s3://nextgen-dmac/kerchunk/hrrr_subhourly.json`
gcs: `gs://squall-hrrr/hrrr_subhourly.json`

The s3 version has references to chunks in s3, the gcs version has references to chunks in gcs.

The notebook with the comparison is available here: https://github.com/mpiannucci/ocean-notebooks/blob/main/hrrr_timeseries.ipynb

The difference in proccessing time is staggering. There are 72 references per variable, so we are loading time, u, and v, then doing compute. That is the time to download and transform 216 references.

**Load Dataset**
| Filesystem      | Time (s) |
| ----------- | ----------- |
| S3      | 0.5       |
| GCS     | 6.9       |

**Extract Wind Speed and Direction Timeseries**

| Filesystem      | Time (s) |
| ----------- | ----------- |
| S3      | 2.7      |
| GCS     | 48       |

However, when I cat the chunks and json there is no difference in speed. 

**Note** I am using my custom grib parser, but tested the same datasets with the cfgrib codec and had the same results. 

Thanks for any help!

**Comment by mpiannucci** [comment]
cfgrib codec versions: 

s3: `s3://nextgen-dmac/kerchunk/hrrr_subhourly_cfgrib.json`

gcs: `gcs://squall-hrrr/hrrr_subhourly_cfgrib.json`

**Comment by martindurant** [comment]
Where do I get the "gribberish" codec?

**Reply by mpiannucci** [reply]
pip install git+https://github.com/mpiannucci/gribberish.git#egg=gribberish&subdirectory=python

**Comment by martindurant** [comment]
I got for GCS:
```
In [1]: import fsspec
In [2]: import xarray as xr
In [3]: %time fs = fsspec.filesystem("reference", fo="gcs://squall-hrrr/hrrr_subhourly_cfgrib.json")
Wall time: 952 ms
In [4]: %time ds = xr.open_dataset(fs.get_mapper(), engine="zarr")
Wall time: 699 ms
```
(where initial connection contains extra latency for auth and ssl negotiation)

for S3:
```
In [1]: import fsspec
In [2]: import xarray as xr
In [3]: %time fs = fsspec.filesystem("reference", fo="s3://nextgen-dmac/kerchunk/hrrr_subhourly_cfgrib.json", target_options={"anon": True})
Wall time: 620 ms
In [4]: %time ds = xr.open_dataset(fs.get_mapper(), engine="zarr")
Wall time: 907 ms
```
.. so, the same?

**Reply by mpiannucci** [reply]
That is suprising. Did you get the same when you download a timeseries subset?

**Reply by martindurant** [reply]
No, I didn't get that far

**Comment by emfdavid** [comment]
I was able to run the notebook from GCS (us-central1)
Here are the timings I see for opening the datasets and extracting the timeseries
<img width="1536" alt="Screenshot 2023-11-17 at 1 05 36 PM" src="https://github.com/fsspec/kerchunk/assets/84335963/4f28aa88-7bcf-4426-a4f4-ec2a0db14187">

**Reply by emfdavid** [reply]
Maybe AWS doesn't like traffic to/from GCS 😆

**Reply by emfdavid** [reply]
Still no where near as fast as the min time you saw though @mpiannucci - even for reading GCS from GCP.

**Reply by mpiannucci** [reply]
Seems wild that i could get it faster from my house haha Are blocks being cached somehow without knowing?

**Reply by martindurant** [reply]
You may wish to turn on the loggers "s3fs", "gcsfs", "fsspec.reference" e.g., 
```
fsspec.utils.setup_logging(logger_name=..)
```

**Reply by martindurant** [reply]
> Are blocks being cached somehow without knowing?

Perhaps you are reusing the filesystem instance? You might want to pepper `skip_instance_cache` in various places

**Reply by mpiannucci** [reply]
Using `skip_instance_cache=True` with both, here are the differences for me running locally:  
<img width="386" alt="Screenshot 2023-11-17 at 1 24 02 PM" src="https://github.com/fsspec/kerchunk/assets/1832671/5742949d-cc54-4a0d-896c-cdae87eda907">
<img width="382" alt="Screenshot 2023-11-17 at 1 24 47 PM" src="https://github.com/fsspec/kerchunk/assets/1832671/526c74b2-4c29-4d79-9c95-cc2aab822311">

**Reply by mpiannucci** [reply]
OKay it was authentication issue. User error. THanks for all the help debugging

**Reply by emfdavid** [reply]
I missed the punch line - what exactly was the issue - no snafu is too dumb for me to hit again on my own I promise!

**Reply by mpiannucci** [reply]
AWS wasn't throwing error on cred error and instead just skipped the chunks, thus the faster chunk fetching times being way too fast

**Reply by martindurant** [reply]
In PR currently, s3fs will "fallback" and try anon access on credentials error. It will work for no-credentials, but maybe not expired-credentials https://github.com/fsspec/s3fs/pull/823

**Reply by mpiannucci** [reply]
Great, that would have cured this so thankful you have a PR for it in the future! Thanks again

**Reply by emfdavid** [reply]
Thanks for working on the optimization @mpiannucci
